### ISSUE TITLE
wrong alignments with scripts and tail output

### ISSUE THREAD

**Issue by x70b1** [polybar/polybar#1903]
A bug in polybar annoys me for a long time. But I could never reproduce it. But I think I'm on the track now I guess.

The elements on the bar go out of the screen from time to time. The reason seems to be scripts that change their width. I've seen this problem with tail scripts, but I can't say if it also happens with simple scripts. But I never noticed that at simple scripts. Also it seems that you need more than one tail scripts. It happens at a group of right aligned modules. Its hard to reproduce.

Apparently the alignment is calculated too rarely or something. I wrote a test script to reproduce it.

**Expected behavior:**

A wondeful `polybar`.

**Actual behavior:**

Everything is ugly and messed up.

**To Reproduce**
```sh
#!/bin/sh

while [ true ]; do
    echo "X X X X X"
    sleep 3

    echo "X"
    sleep 3
done
```
Ugly:
![2019-10-17-21:54:00-screenshot](https://user-images.githubusercontent.com/10016267/67043086-88a55300-f129-11e9-86c2-dbc08cea03ab.png)
Wonderful
![2019-10-17-21:54:03-screenshot](https://user-images.githubusercontent.com/10016267/67043134-abd00280-f129-11e9-85d6-d86c1c9ac5d5.png)

```
polybar 3.4.0-20-g4ab251f3

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -Wall -Wextra -Wpedantic -O3 -DNDEBUG
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now 
```

**Comment by patrick96** [comment]
That is very interesting. This could also be the cause for many of the "tray overlapping" issues, maybe it's not a tray issue but just that the script modules shift the bar over.

**Comment by patrick96** [comment]
How often does this happen? I am not able to reproduce this at all.

Also strange that the bar shift to the right when there is *less* text.

And also the alignment is calculated everytime the bar is rerendered. The x coordinate of the right block is either the bar width minus the block length or the x coordinate of the rightmost edge of the center block plus a gap of 20 pixels, whichever is greater. Generally the first one is greater since that means that the right block fully fits into the bar

I have looked through the code and I couldn't find any apparent errors, if you can reliably reproduce this and would like to play around with `gdb`, have a look at what happens in this code section:
https://github.com/polybar/polybar/blob/4ab251f33ca5cec1f8de8468ee37e36a59eaed09/src/components/renderer.cpp#L456-L461

Here `std::max` should return a value that is too large and pushes the right block too far out.

**Comment by x70b1** [comment]
It looks like #1268 is the same issue.
It's very rare. I couldn't see it again until now.

**Comment by patrick96** [comment]
That's possible. I wasn't able to reproduce that issue either.

**Comment by patrick-motard** [comment]
Pretty sure I'm seeing the same issue.
![2019-12-14_22_07_38_532x75](https://user-images.githubusercontent.com/6657259/70857962-287b2500-1ebe-11ea-9907-709829b01f18.png)
Settings:
```
[bar/main.top.right]
; require monitor to be connected to load bar
monitor-strict = true
inherit = bar/base
monitor = ${env:MONITOR_MAIN}
modules-right =  alsa-switch pulseaudio-tail time calendar dropbox-watcher
; dropbox-watcher
; alsa-switch pulseaudio-tail
modules-left = 
modules-center =
border = 2
background = ${colors.nord0}
width = 500
bottom = false 
; tray-position = none
height = 3%
padding = 0
override-redirect = true
; offset-x = 84.5%
offset-x = 74.5%
offset-y = 20
border-size = 4
border-color = ${colors.nord2}
tray-position = right
tray-background = ${colors.nord0}
```
"dropbox-watcher", "alsa-switch", and "pulseaudio-tail" are all scripts i've written, whose output changes dynamically.
Here is what "dropbox-watcher"'s script is:
```
#!/usr/bin/env bash

case "$1" in
    --toggle)
        if [ "$(pgrep dropbox)" ]; then
            pkill -f dropbox
        else
            dropbox &
        fi
        ;;
    *)
        if [ "$(pgrep dropbox)" ]; then
            echo " %{F#A3BE8C}%{F-} "
        else
            echo " %{F#EBCB8B}%{F-} "
        fi
        ;;
esac
```
I've confirmed the overlap happens no matter which of the custom scripts i put in the bar.
If i put the tray on the left, there are no overlap issues:
![2019-12-14_22_09_51_549x82](https://user-images.githubusercontent.com/6657259/70857981-7b54dc80-1ebe-11ea-8874-337ea3c21292.png)

Also, if i remove my custom scripts, and instead leave the built in modules, and place the tray on the right again, there is no overlap:
```
modules-right =  time calendar
tray-position = right
```
![2019-12-14_22_12_15_524x65](https://user-images.githubusercontent.com/6657259/70857998-cb33a380-1ebe-11ea-9e5b-594d0210efb1.png)
Another interesting find is that I'm able to avoid the overlap issue if i make the bar MUCH wider than it needs to be. You can see in the screenshot below that the gap between the left side of the bar and the volume icon is large. I'd like it to be flush. Not possible though, due to the overlap occuring:
```
width = 600
; instead of the original 500
```
![2019-12-14_22_14_10_623x63](https://user-images.githubusercontent.com/6657259/70858017-0f26a880-1ebf-11ea-999b-00faf5f25ade.png)

Full transparency, I'm doing some hacky things to get 3 bars in parallel. With the source code I'll link below, I'm able to achieve the following:

![2019-12-14_22_15_15_2554x289](https://user-images.githubusercontent.com/6657259/70858026-4301ce00-1ebf-11ea-96e7-e9355a459b76.png)

Polybar configs: https://github.com/patrick-motard/dotfiles/tree/master/.config/polybar/themes/nord

**Comment by patrick-motard** [comment]
This issue does not exist in a slightly different configuration. In my [nord-arrow theme](https://github.com/patrick-motard/dotfiles/blob/master/.config/polybar/themes/nord-arrow/config), the bottom bar doesn't have this issue. It uses the same underlying modules as my [nord theme](https://github.com/patrick-motard/dotfiles/blob/master/.config/polybar/themes/nord/config)

![2019-12-14_22_36_49_2558x63](https://user-images.githubusercontent.com/6657259/70858177-39c63080-1ec2-11ea-8ae5-22d37fee3cbb.png)

Here's the configuration for that bar: https://github.com/patrick-motard/dotfiles/blob/master/.config/polybar/themes/nord-arrow/config#L44

**Comment by patrick96** [comment]
@patrick-motard Thanks for the insight. What you describe is consistent with what I have experienced. With the tray on the right, no center modules and `fixed-center = true` or not set at all, there is some overlap once the modules on the right go over the middle of the bar. My fix in #1936 should fix that.

**Comment by patrick96** [comment]
@x70b1 I am now pretty sure that this and #1268 are the same underlying issue. However I was so far not able to reproduce the shifting of the modules on the right without having a tray on the right or `border-right-size` set. Do you have set a border on the right?

**Comment by x70b1** [comment]
@patrick96 sorry for the late response.
No border size in my config. :cry: 

I use a module margin for everything:
```
module-margin-right = 5
module-margin-left = 5
```

**Comment by patrick96** [comment]
> I use a module margin for everything:

That shouldn't have any effect. Nonetheless, do you still get the wrong alignment with my PR?

**Comment by x70b1** [comment]
I cant reproduce the issue at the moment. 
But as you find out something and fixed it, I would be ok with closing the issues. 

If it comes back, we can reopen the case.

**Comment by patrick96** [comment]
Was finally able to reproduce this reliably. The reason this is so tricky to track down without border-right or the tray is because the left edge of the modules on the right has to be in a 20pixel wide zone directly to the right of the bar center to actually experience this. And it is really only noticeable if it's even closer.

Those 20 pixels is the gap that is inserted between modules-left, modules-center and modules-right and those were not taken into account in the alignment calculations. Also, there *has* to be a module on the left, otherwise no gap is added and the bug isn't triggered.

My fix in #1936 fixes this because it completely ignores the center block when it's empty.

If anyone is interested, the following is my minimal config with which I was able to reproduce this:

```dosini
[bar/example]
modules-left = filler
modules-right = text

font-0 = Dejavu Sans Mono:size=12

dpi = 96

[module/filler]
type = custom/text
content = X

[module/text]
type = custom/text
content = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXText
```

I have a 1920x1080 monitor, you will likely only be able to reproduce this if your monitor settings match mine exactly. Otherwise if your monitor is larger, add X's to the left of the `text` module on the right until the module almost reaches the center, around there the shifting should occur.

**Comment by x70b1** [comment]
Wonderful! :heart:
